src/drivers/net/marvell/atl_hw.c

```diff
@@ -0,0 +1,317 @@
+		      ATL_TX_CTRL);
+	ATL_WRITE_REG(ATL_READ_REG(ATL_MAC_PHY_CTRL) &
+		      ~ATL_MAC_PHY_CTRL_RST_DIS, ATL_MAC_PHY_CTRL);
+
+	ATL_WRITE_REG((ATL_READ_REG(ATL_GLB_STD_CTRL) &
+		      ~ATL_GLB_CTRL_RST_DIS) | 0x8000, ATL_GLB_STD_CTRL);
+
+	ATL_WRITE_REG(0x40e0, ATL_GLB_CTRL2);
+
+	/* Wait for RBL boot */
+	for (k = 0; k < 1000; k++) {
+		rbl_status = ATL_READ_REG(ATL_MPI_BOOT_EXIT_CODE) & 0xFFFF;
+		if (rbl_status && rbl_status != 0xDEAD)
+			break;
+		mdelay(10);
+	}
+	if (!rbl_status || rbl_status == 0xDEAD) {
+		DBG("RBL Restart failed");
+		return -EIO;
+	}
+
+	if (rbl_status == 0xF1A7)
+		return -ENOTSUP;
+
+	for (k = 0; k < 1000; k++) {
+		u32 fw_state = ATL_READ_REG(ATL_FW_VER);
+
+		if (fw_state)
+			break;
+		mdelay(10);
+	}
+	if (k == 1000) {
+		DBG("FW kickstart failed\n");
+		return -EIO;
+	}
+	/* Old FW requires fixed delay after init */
+	mdelay(15);
+
+	return 0;
+}
+
+int atl_hw_reset(struct atl_nic *nic)
+{
+	uint32_t boot_exit_code = 0;
+	uint32_t k;
+	int rbl_enabled;
+	uint32_t fw_ver;
+	uint32_t sem_timeout;
+
+	for (k = 0; k < 1000; ++k) {
+		uint32_t flb_status = ATL_READ_REG(ATL_MPI_DAISY_CHAIN_STS);
+		boot_exit_code = ATL_READ_REG(ATL_MPI_BOOT_EXIT_CODE);
+		if (flb_status != 0x06000000 || boot_exit_code != 0)
+			break;
+	}
+
+	if (k == 1000) {
+		DBG("Neither RBL nor FLB firmware started\n");
+		return -ENOTSUP;
+	}
+
+	rbl_enabled = (boot_exit_code != 0);
+
+	fw_ver = ATL_READ_REG(ATL_FW_VER);
+	if (((fw_ver >> 24) & 0xFF) >= 4) {
+		sem_timeout = ATL_READ_REG(ATL_SEM_TIMEOUT);
+		if (sem_timeout > 3000)
+			sem_timeout = 3000;
+
+		for (k = 0; k < sem_timeout; ++k) {
+			if (ATL_READ_REG(ATL_GLB_MCP_SEM4))
+				break;
+
+			mdelay(1);
+		}
+		for (k = 0; k < sem_timeout; ++k) {
+			if (ATL_READ_REG(ATL_GLB_MCP_SEM5))
+				break;
+
+			mdelay(1);
+		}
+	}
+
+
+	if (rbl_enabled)
+		return atl_hw_reset_rbl_(nic);
+	else
+		return atl_hw_reset_flb_(nic);
+}
+
+int atl_hw_start(struct atl_nic *nic)
+{
+	ATL_WRITE_REG(ATL_LINK_ADV_AUTONEG, ATL_LINK_ADV);
+	return 0;
+}
+
+int atl_hw_stop(struct atl_nic *nic)
+{
+	ATL_WRITE_REG(0x0, ATL_LINK_ADV);
+	return 0;
+}
+
+int atl_hw_get_link(struct atl_nic *nic)
+{
+	return (ATL_READ_REG(ATL_LINK_ST) & ATL_LINK_ADV_AUTONEG) != 0;
+}
+
+int atl_hw_read_mem(struct atl_nic *nic, uint32_t addr, uint32_t *buffer,
+		    uint32_t size)
+{
+	uint32_t i;
+	DBG("AQUANTIA: atl_hw_read_mem\n");
+
+	for (i = 0; i < 100; ++i) {
+		if (ATL_READ_REG(ATL_SEM_RAM))
+			break;
+		mdelay(1);
+	}
+	if (i == 100)
+		goto err;
+
+	ATL_WRITE_REG(addr, ATL_MBOX_CTRL3);
+
+	for (i = 0; i < size; ++i, addr += 4) {
+		uint32_t j;
+
+		ATL_WRITE_REG(0x8000, ATL_MBOX_CTRL1);
+		for (j = 0; j < 10000; ++j) {
+			if (ATL_READ_REG(ATL_MBOX_CTRL3) != addr)
+				break;
+			udelay(10);
+		}
+		if (j == 10000)
+			goto err;
+
+		buffer[i] = ATL_READ_REG(ATL_MBOX_CTRL5);
+	}
+
+	ATL_WRITE_REG(1, ATL_SEM_RAM);
+
+	return 0;
+err:
+	DBG("AQUANTIA: download_dwords error\n");
```

Comment:
Please include meaningful error messages.  You have what appears to be two unrelated possible causes for this error: they should not produce the same message.

This comment applies throughout the driver.